### PR TITLE
Fix x86 linux tests

### PR DESCRIPTION
**Pull request by t-mustafin**
This PR contains commits which fixes build and runtime tests bugs on linux x86 platform. A majority of them fixes calling convention issues.

### CODE REVIEW

**Comment by t-mustafin** [comment]
cc @jkotas @janvorli

File: src/coreclr/vm/i386/umthunkstub.S [hunk 1]
```diff
@@ -9,14 +9,23 @@
...
 NESTED_ENTRY TheUMEntryPrestub, _TEXT, UnhandledExceptionHandlerUnix
-#define STK_ALIGN_PADDING 8
-    sub     esp, STK_ALIGN_PADDING
+
+// Calculate stack alignment
+    push    ebx
+#define STACK_ALIGN_MASK 0xf
+    mov     ebx, esp
+    sub     ebx, 4
+    and     ebx, STACK_ALIGN_MASK
```
Comment (jkotas) [inline src/coreclr/vm/i386/umthunkstub.S:18]: This should not be needed. The caller should guarantee the right aligment.

File: src/coreclr/jit/importer.cpp [hunk 2]
```diff
@@ -8219,6 +8222,13 @@ var_types Compiler::impImportCall(OPCODE                  opcode,
         szCanTailCallFailReason = "Caller is varargs";
     }
 #endif // FEATURE_FIXED_OUT_ARGS
+#if UNIX_X86_ABI
+    else
+    {
+        canTailCall             = false;
+        szCanTailCallFailReason = "Platform is Unix x86";
+    }
+#endif // UNIX_X86_ABI
```
Comment (jakobbotsch) [inline src/coreclr/jit/importer.cpp:8231]: What problem did you run into with Unix x86 tailcalls? This is probably not right as it will deny even helper-based tailcalls that should just work on x86.
I think instead you should reject using the Windows x86 tailcall helper unconditionally in this function:
https://github.com/dotnet/runtime/blob/a579575853cea23aa1b6b6b8d3ef978aa9321e16/src/coreclr/jit/morph.cpp#L18172-L18191

File: src/coreclr/jit/morph.cpp [hunk 3]
```diff
@@ -2805,17 +2805,8 @@ void Compiler::fgInitArgInfo(GenTreeCall* call)
...
-        GenTreeCall::Use** insertionPoint = &call->gtCallArgs;
-        for (; *insertionPoint != nullptr; insertionPoint = &((*insertionPoint)->NextRef()))
-        {
-        }
-        *insertionPoint = gtNewCallArgs(arg);
-#else  // !defined(TARGET_X86)
-        // All other architectures pass the cookie in a register.
+        // All architectures pass the cookie in a register.
         call->gtCallArgs = gtPrependNewCallArg(arg, call->gtCallArgs);
-#endif // defined(TARGET_X86)
```
Comment (jakobbotsch) [inline src/coreclr/jit/morph.cpp:2818]: What is the reason for this change? Did you measure impact of it? Since `ebx` is callee saved this might be pretty expensive to use as the new register given how few callee-saved registers are available on x86.

File: src/coreclr/vm/i386/jithelp.S [hunk 4]
```diff
@@ -387,16 +388,11 @@ C_FUNC(JIT_WriteBarrierEAX_Loc):
...
-.att_syntax
-    addl     $_GLOBAL_OFFSET_TABLE_+(2b-1b), %eax
-.intel_syntax noprefix
-    mov     eax, dword ptr [eax + C_FUNC(JIT_WriteBarrierEAX_Loc)@GOT]
+// Place JIT_WriteBarrierEAX address into stack to call it via ret instuction.
+    push eax
+    mov eax, dword ptr [C_FUNC(JIT_WriteBarrierEAX_Loc)]
     xchg    eax, dword ptr [esp]
+// Real return address is still on the stack at esp + 4.
     ret
```
Comment (jakobbotsch) [inline src/coreclr/vm/i386/jithelp.S:396]: Are we sure that the above removed instructions are not trying to keep the hardware's call-ret stack balanced?

File: src/coreclr/jit/morph.cpp (hunk 3 above)
Comment (jkotas) [reply src/coreclr/jit/morph.cpp:2818]: calli using this helper is slow path for multiple difference reasons. We are optimizing for simplicity here. This change is deleting number of instructions from the helper that I expect will more than compensate for the ebx use here.

To make this go fast, we would need to get rid of this helper completely and call the marshaling stub directly from here. It would be more involved change, beyond the scope of this PR.

File: src/coreclr/inc/corinfo.h [hunk 5]
```diff
@@ -720,6 +720,31 @@ enum class CorInfoCallConvExtension
 };
 
 #ifdef TARGET_X86
+inline const char *GetCallConvName(CorInfoCallConvExtension callConv)
```
Comment (BruceForstall) [inline src/coreclr/inc/corinfo.h:723]: This code shouldn't be in the "interface" header function. It looks like it's only used by debug code in the JIT, so should be defined in the JIT, only under `#ifdef DEBUG`

File: src/coreclr/jit/reglist.h [hunk 6]
```diff
@@ -9,7 +9,7 @@
 
 // The "regList" type is a small set of registerse
 #ifdef TARGET_X86
-typedef TinyArray<unsigned short, regNumber, REGNUM_BITS> regList;
+typedef TinyArray<unsigned int, regNumber, REGNUM_BITS> regList;
```
Comment (BruceForstall) [inline src/coreclr/jit/reglist.h:12]: Why this change?

**Review by BruceForstall — CHANGES_REQUESTED** [review]

File: src/coreclr/jit/morph.cpp (hunk 3 above)
Comment (jakobbotsch) [reply src/coreclr/jit/morph.cpp:2818]: My main worry was that reserving ebx can bias the register allocator into spilling more variables that are live across calls which will affect even the paths that do not run the expensive calli. However I checked win x86 diffs and there is only a single non-test diff in libraries with this change (https://github.com/dotnet/runtime/blob/cffaa78235ea93d5e3eeb56956579df503e11250/src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Internal/MsQuicApi.cs#L144-L146) so it seems fine to me.

File: src/coreclr/vm/i386/jithelp.S (hunk 4 above)
Comment (jakobbotsch) [reply src/coreclr/vm/i386/jithelp.S:396]: If no IP-relative computation is needed I think the entire function can just be
```asm
mov eax, edx
mov edx, ecx
jmp dword ptr [C_FUNC(JIT_WriteBarrierEAX_Loc)]
```

File: src/coreclr/jit/reglist.h (hunk 6 above)
Comment (t-mustafin) [reply src/coreclr/jit/reglist.h:12]: It caused by `ebx` usage for `VASigCookie` passing to `GenericPInvokeCalliHelper`. `ebx` is a fourth register (eax, ecx, edx, ebx) and `REGNUM_BITS = 6`: it leads to overflow of `unsigned short` type. This overflow appears via assertion:
https://github.com/dotnet/runtime/blob/9703660baa08914773b26e413e361c8ce04e6d94/src/coreclr/jit/gentree.cpp#L12334
Right part was 0 for `ebx` due to this overflow.

File: src/coreclr/inc/corinfo.h (hunk 5 above)
Comment (t-mustafin) [reply src/coreclr/inc/corinfo.h]: Thanks!

File: src/coreclr/jit/importer.cpp (hunk 2 above)
Comment (t-mustafin) [reply src/coreclr/jit/importer.cpp]: It is a fix of about 130 failing tests with message example:
```
FAILED   - [   2s]JIT/Directed/pinvoke/tail/tail.sh

               BEGIN EXECUTION
               /home/tmustafin/dotnet/x86_release_test/coreroot//corerun tail.dll ''
               Unhandled exception. System.InvalidProgramException: Common Language Runtime detected an invalid program.
                  at JitTest.Test.__DestroyMenu(UIntPtr hMenu)
                  at JitTest.Test.Main() in /home/tmustafin/dotnet/runtime/src/tests/JIT/Directed/pinvoke/tail.il:line 105
               ./tail.sh: line 375: 30509 Aborted                 (core dumped) $LAUNCHER $ExePath "${CLRTestExecutionArguments[@]}"
               Expected: 100
               Actual: 134
               END EXECUTION - FAILED

```
However, fix in `fgCanTailCallViaJitHelper` works better for couple of tests. Thanks!

Comment (jakobbotsch) [reply src/coreclr/jit/importer.cpp]: That's weird. Is it possible for you to retrieve a JIT dump from one of the failing tail call test cases?

File: src/coreclr/vm/i386/jithelp.S (hunk 4 above)
Comment (t-mustafin) [reply src/coreclr/vm/i386/jithelp.S:396]: > Are we sure that the above removed instructions are not trying to keep the hardware's call-ret stack balanced?

Maybe I do not clearly understand the question. I am sure that the above removed instructions keep the hardware's call-ret info on stack balanced. Removed instructions just calculate wrong address.




> If no IP-relative computation is needed I think the entire function can just be
> 
> ```assembly
> mov eax, edx
> mov edx, ecx
> jmp dword ptr [C_FUNC(JIT_WriteBarrierEAX_Loc)]
> ```

Unfortunately this code jumps to `JIT_WriteBarrierEAX_Loc`, not to the address stored in `JIT_WriteBarrierEAX_Loc`:
```gdb
=> 0xf7606c71 <JIT_WriteBarrier_Callable>:      mov    eax,edx
   0xf7606c73 <JIT_WriteBarrier_Callable+2>:    mov    edx,ecx
   0xf7606c75 <JIT_WriteBarrier_Callable+4>:    jmp    0xf7ad49b4
gdb$ disassemble 0xf7ad49b4
Dump of assembler code for function JIT_WriteBarrierEAX_Loc:
   0xf7ad49b4:  enter  0x6075,0xf7
End of assembler dump.
gdb$ disassemble 0xf76075c8
Dump of assembler code for function JIT_WriteBarrierEAX:
   0xf76075c8 <+0>:     mov    DWORD PTR [edx],eax
   0xf76075ca <+2>:     cmp    eax,0xf1a0000c
   0xf76075d0 <+8>:     jb     0xf76075df <JIT_WriteBarrierEAX+23>
   0xf76075d2 <+10>:    shr    edx,0xa
   0xf76075d5 <+13>:    nop
   0xf76075d6 <+14>:    cmp    BYTE PTR [edx-0xa4007dc],0xff
   0xf76075dd <+21>:    jne    0xf76075e2 <JIT_WriteBarrierEAX+26>
   0xf76075df <+23>:    ret    
   0xf76075e0 <+24>:    nop
   0xf76075e1 <+25>:    nop
   0xf76075e2 <+26>:    mov    BYTE PTR [edx-0xa4007dc],0xff
   0xf76075e9 <+33>:    ret 
```

File: src/coreclr/jit/reglist.h (hunk 6 above)
Comment (BruceForstall) [reply src/coreclr/jit/reglist.h:12]: Ok. Seems like some assert in tinyarray.h should have fired with this overflow.

File: src/coreclr/vm/i386/jithelp.S (hunk 4 above)
Comment (jakobbotsch) [reply src/coreclr/vm/i386/jithelp.S:396]: > Maybe I do not clearly understand the question. I am sure that the above removed instructions keep the hardware's call-ret info on stack balanced. Removed instructions just calculate wrong address.

The hardware tracks `call`/`ret` pairs to be able to branch predict `ret` instructions better. It means that after this change the hardware might mispredict a lot of the upcoming `ret` instructions.

> Unfortunately this code jumps to JIT_WriteBarrierEAX_Loc, not to the address stored in JIT_WriteBarrierEAX_Loc:

That's odd. Maybe try AT&T syntax? It should assemble to `ff 25 imm32`.

Comment (jakobbotsch) [reply src/coreclr/vm/i386/jithelp.S:396]: I can reproduce it with `clang-9 foo.S` locally. Using AT&T syntax instead works and assembles correctly:
```asm
.att_syntax
    jmpl *C_FUNC(JIT_WriteBarrierEAX_loc)
```

Comment (jakobbotsch) [reply src/coreclr/vm/i386/jithelp.S:396]: https://bugs.llvm.org/show_bug.cgi?id=20511

File: src/coreclr/jit/importer.cpp (hunk 2 above)
Comment (t-mustafin) [reply src/coreclr/jit/importer.cpp]: It is an example dump: https://gist.github.com/t-mustafin/2a60e785b148e7f496347979ea2bc26e.
Collected by:
```
DOTNET_JitDump="__DestroyMenu" ./corerun ~/dotnet/rebased2_x86_release_test/JIT/Directed/pinvoke/tail/tail.dll > ~/tasks/88_x86_helloworld/test_phase/tailcall_feedback/tail.dump
Unhandled exception. System.InvalidProgramException: Common Language Runtime detected an invalid program.
   at JitTest.Test.__DestroyMenu(UIntPtr hMenu)
   at JitTest.Test.Main() in /home/tmustafin/dotnet/runtime/src/tests/JIT/Directed/pinvoke/tail.il:line 105
Aborted (core dumped)
```

Comment (jakobbotsch) [reply src/coreclr/jit/importer.cpp]: The dump seems to suggest that `fgCanTailCallViaJitHelper` returned true since it is still being morphed as a tail call via JIT helper:
```scala
fgMorphTailCallViaJitHelper (before):
               [000005] --CXG-------              *  CALL ind  void  
               [000000] ------------ arg0         +--*  LCL_VAR   int    V00 arg0         
               [000004] ------------ calli tgt    \--*  LCL_VAR   int    V01 tmp0         
fgMorphTailCallViaJitHelper (after):
               [000005] --CXG-------              *  CALL ind  void  
               [000000] ------------ arg0         +--*  LCL_VAR   int    V00 arg0         
               [000007] ------------ arg1         +--*  CNS_INT   int    0
               [000008] ------------ arg2         +--*  CNS_INT   int    9
               [000009] ------------ arg3         +--*  CNS_INT   int    8
               [000010] ------------ arg4         +--*  CNS_INT   int    7
               [000004] ------------ calli tgt    \--*  LCL_VAR   int    V01 tmp0         
```
Are you sure this is with your newest changes?

File: src/coreclr/jit/reglist.h (hunk 6 above)
Comment (t-mustafin) [reply src/coreclr/jit/reglist.h:12]: Thanks, I have added a fix into assert in tinyarray.h.

File: src/coreclr/vm/i386/asmhelpers.S [hunk 7]
```diff
@@ -620,60 +620,33 @@ NESTED_END VarargPInvokeStub, _TEXT
...
-    // ------------------------
+    // | CALLI target address | ESP + 0
+    // +----------------------+
+
+    // load EAX with the target address, store on stack stub entry point
+    xchg    eax, [esp]
 
-    // CALLI target address is in EAX
+    // call stub
     ret
```
Comment (t-mustafin) [inline src/coreclr/vm/i386/asmhelpers.S:650]: @jakobbotsch Should I change this `ret` instruction to `jmp` also?

File: src/coreclr/vm/i386/asmhelpers.asm [hunk 8]
```diff
@@ -1142,60 +1142,33 @@ _VarargPInvokeStub@0 endp
...
-    ; ------------------------
+    ; | CALLI target address | ESP + 0
+    ; +----------------------+
+
+    ; load EAX with the target address, store on stack stub entry point
+    xchg    eax, [esp]
 
-    ; CALLI target address is in EAX
+    ; call stub
     ret
```
Comment (t-mustafin) [inline src/coreclr/vm/i386/asmhelpers.asm:1172]: @jakobbotsch Should I change this `ret` instruction to `jmp` also?

File: src/coreclr/jit/reglist.h (hunk 6 above)
Comment (t-mustafin) [reply src/coreclr/jit/reglist.h:12]: Thanks, I've fixed assert in tinyarray.h.

File: src/coreclr/jit/importer.cpp (hunk 2 above)
Comment (t-mustafin) [reply src/coreclr/jit/importer.cpp]: It was on state before any tail call patch. With newest changes no tail call problem occures.

> However, fix in `fgCanTailCallViaJitHelper` works better for couple of tests. Thanks!
I mean that patch in `fgCanTailCallViaJitHelper` fixes all test which was fixed by previous patch + 2 more.

File: src/coreclr/vm/i386/jithelp.S (hunk 4 above)
Comment (t-mustafin) [reply src/coreclr/vm/i386/jithelp.S]: Thanks, AT&T syntax works.

File: src/coreclr/vm/i386/asmhelpers.asm (hunk 8 above)
Comment (jakobbotsch) [reply src/coreclr/vm/i386/asmhelpers.asm:1172]: I think it's not important for these helpers for the reasons @jkotas mentioned. The write barriers are much hotter (though not sure about the callable version).
But if it makes the stubs simpler then go for it.

File: src/coreclr/jit/importer.cpp (hunk 2 above)
Comment (jakobbotsch) [reply src/coreclr/jit/importer.cpp]: Ah, my misunderstanding. Great to hear that it works.

File: src/coreclr/vm/i386/asmhelpers.asm (hunk 8 above)
Comment (jkotas) [reply src/coreclr/vm/i386/asmhelpers.asm:1172]: It would be nice to change this to `jmp` also. It should be easy to do now that there is an extra register to use.

Comment (t-mustafin) [reply src/coreclr/vm/i386/asmhelpers.asm]: Thanks, done.

**Review by jakobbotsch — APPROVED** [review]

**Review by jkotas — APPROVED** [review]

File: src/coreclr/vm/i386/jithelp.S (hunk 4 above)
Comment (janvorli) [reply src/coreclr/vm/i386/jithelp.S]: IP relative addressing is a requirement on Linux. While it works without it, it requires relocations when loading the code and LLVM linker even refuses to link such code with our current options. I had to fix it recently for Linux ARM.

Comment (t-mustafin) [reply src/coreclr/vm/i386/jithelp.S]: Seems `eax` dereferencing was missed in original IP-relative addressing. Updated a PR, @janvorli please take a look.

Comment (t-mustafin) [reply src/coreclr/vm/i386/jithelp.S]: Hmm, with GOT we do not need `JIT_WriteBarrierEAX_Loc` at all. The code below works properly:
```asm
    mov eax, edx
    mov edx, ecx
    push    eax
    call    1f
1:
    pop     eax
2:
.att_syntax
    addl     $_GLOBAL_OFFSET_TABLE_+(2b-1b), %eax
.intel_syntax noprefix
    mov     eax, dword ptr [eax + C_FUNC(JIT_WriteBarrierEAX)@GOT]
    xchg    eax, dword ptr [esp]
    ret
```

Comment (janvorli) [reply src/coreclr/vm/i386/jithelp.S]: I don't see how this would work with the W^X enabled. We don't want to call the JIT_WriteBarrierEAX in the executable, but rather its copy that is placed at a random memory location (in a dynamically allocated memory page).

Comment (t-mustafin) [reply src/coreclr/vm/i386/jithelp.S]: > I don't see how this would work with the W^X enabled. We don't want to call the JIT_WriteBarrierEAX in the executable, but rather its copy that is placed at a random memory location (in a dynamically allocated memory page).

Would current PR variant with JIT_WriteBarrierEAX_Loc dereferencing work with W^X enabled?

Comment (janvorli) [reply src/coreclr/vm/i386/jithelp.S]: Yes, if that works with W^X disabled, it would work with it enabled too. The JIT_WriteBarrierEAX_Loc points to the JIT_WriteBarrierEAX with W^X disabled and to its copy with W^X enabled.

**Review by BruceForstall — APPROVED** [review]
